src/Compilers/CSharp/Portable/Parser/LanguageParser.cs

```diff
@@ -1962,6 +1963,151 @@ static TypeDeclarationSyntax constructTypeDeclaration(ContextAwareSyntax syntaxF
             }
         }
 
+        private MemberDeclarationSyntax ParseExtensionContainer(SyntaxList<AttributeListSyntax> attributes, SyntaxListBuilder modifiers)
+        {
+            Debug.Assert(!IsInAsync);
+
+            cancellationToken.ThrowIfCancellationRequested();
+
+            Debug.Assert(this.CurrentToken.ContextualKind == SyntaxKind.ExtensionKeyword);
+            SyntaxToken extensionKeyword = this.EatContextualToken(SyntaxKind.ExtensionKeyword);
+
+            var outerSaveTerm = _termState;
+            _termState |= TerminatorState.IsPossibleAggregateClauseStartOrStop; // TODO2 add test affected
+
+            TypeParameterListSyntax typeParameters = this.ParseTypeParameterList();
+            ReceiverParameterSyntax receiverParameter = parseReceiverParameter(out var openParenToken, out var closeParenToken);
+
+            // Parse extension container body
+            bool parseMembers = true;
+            SyntaxListBuilder<TypeParameterConstraintClauseSyntax> constraints = default;
+            SyntaxListBuilder<MemberDeclarationSyntax> members = default;
+            try
+            {
+                if (this.CurrentToken.ContextualKind == SyntaxKind.WhereKeyword)
+                {
+                    _termState |= TerminatorState.IsEndOfRecordOrClassOrStructOrInterfaceSignature;
+                    constraints = _pool.Allocate<TypeParameterConstraintClauseSyntax>();
+                    this.ParseTypeParameterConstraintClauses(constraints);
+                }
+
+                _termState = outerSaveTerm;
+
+                SyntaxToken? openBrace = this.EatToken(SyntaxKind.OpenBraceToken);
+                SyntaxToken? closeBrace;
+
+                // ignore members if missing open curly
+                if (openBrace.IsMissing)
+                {
+                    parseMembers = false;
+                }
+
+                if (parseMembers)
+                {
+                    members = _pool.Allocate<MemberDeclarationSyntax>();
+
+                    while (true)
```

Comment:
afaict, this is a copy of ParseClassOrStructOrInterfaceDeclaration.  Seems like we can just expand that to parse extensions, just with some extension-specific code.  Note: as all teh rest support the primary cosntructor parameter list, this all falls out.   The only difference that i can tell so far is simple that in the case of extensions we want at least one parameter (though this could be a binding-time check), and that we do *not* have an identifier.  All of that seems similar to add to ParseClassOrStructOrInterfaceDeclaration vs this copy :)